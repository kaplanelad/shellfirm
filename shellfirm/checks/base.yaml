- from: base
  test: ":(){ :|: & };:"
  method: Contains
  enable: true
  description: "This short line defines a shell function that creates new copies of itself.\nThe process continually replicates itself, and its copies continually replicate themselves, quickly taking up all your CPU time and memory.\nThis can cause your computer to freeze. Itâ€™s basically a denial-of-service attack."
- from: base
  test: \s*crontab\s+-r
  method: Regex
  enable: true
  description: "You are going to remove your entire table of cron tasks."
- from: base
  test: \s*history.*.[|].*.(bash|sh)
  method: Regex
  enable: true
  description: "You are going to executes every command from the command log that you have already executed."
- from: base
  test: (\.*)reboot[^."']*{0,}
  method: Regex
  enable: true
  description: "You are going to reboot your machine."
- from: base
  test: (\.*)shutdown[^."']*{0,}
  method: Regex
  enable: true
  description: "You are going to shutdown your machine."
