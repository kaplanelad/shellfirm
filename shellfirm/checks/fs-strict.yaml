- from: fs-strict
  test: chmod\s
  description: "Are you sure that you want to change permissions?"
  id: fs-strict:change_permissions
- from: fs-strict
  test: ^rm(?:\s+-?[rRfF]*)?\s+(.*)
  description: "Are you sure you want to continue with deletion?"
  id: fs-strict:any_deletion
  filters:
    IsExists: "1"
- from: fs-strict
  test: rmdir\s
  description: "Are you sure you want to continue with deletion?"
  id: fs-strict:folder_deletion
  filters:
    IsExists: "1"
