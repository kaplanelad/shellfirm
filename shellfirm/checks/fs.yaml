- from: fs
  test: 'rm\s{1,}(?:-R|-r|-fR|-fr|-Rf|-rf|-v|--force|--verbose|--preserve-root)\s*(?:-R|-r|-fR|-fr|-Rf|-rf|-v|--force|--verbose|--preserve-root)?\s*(?:\*|\.{1,}|/)\s*$'
  description: "You are going to delete everything in the path."
  id: fs:recursively_delete
  filters:
    IsExists: "3"
- from: fs
  test: mv\s+(?:-[a-zA-Z]+\s+)?(?:--[a-zA-Z-]+\s+)?(?:[^\s]+\s+)+/dev/null
  description: "The files will be discarded and destroyed."
  id: fs:move_to_dev_null
  filters:
    IsExists: "1"
- from: fs
  test: .*>(.*)
  description: "The above command is used to flush the content of a file."
  filters:
    IsExists: "1"
  id: fs:flush_file_content
- from: fs
  test: chmod\s{1,}(-R|--recursive)\s{1}(\*|\.{2,}|/)
  description: "Change permission to all root files can brake your some thinks like SSH keys."
  id: fs:recursively_chmod
- from: fs
  test: find\s.*-delete
  description: "Did you -delete flag in the wrong order? find -delete going to delete all the file under your current path."
  id: fs:delete_find_files
- from: fs
  test: 'dd\s+.*of=/dev/([hs]d[a-z]|mmcblk[0-9])'
  description: "Are you sure you want to write directly to a block device? This could overwrite your disk."
  id: fs:dd_block_device
- from: fs
  test: 'mkfs(?:\.(?:ext[2-4]|fat|ntfs|xfs|btrfs))?\s+(?:-t\s+\w+\s+)?/dev/([hs]d[a-z][0-9]*|mmcblk[0-9]p?[0-9]*)'
  description: "Are you sure you want to format this device? This will erase all data on it."
  id: fs:mkfs_format
- from: fs
  test: 'parted\s+.*(/dev/([hs]d[a-z]|mmcblk[0-9]))'
  description: "Are you sure you want to modify disk partitions? This could erase all data on the disk."
  id: fs:parted_disk_modify
- from: fs
  test: 'fdisk\s+.*(/dev/([hs]d[a-z]|mmcblk[0-9]))'
  description: "Are you sure you want to modify disk partitions? This could erase all data on the disk."
  id: fs:fdisk_disk_modify
- from: fs
  test: 'sfdisk\s+.*(/dev/([hs]d[a-z]|mmcblk[0-9]))'
  description: "Are you sure you want to modify disk partitions? This could erase all data on the disk."
  id: fs:sfdisk_disk_modify
- from: fs
  test: 'dd\s+.*(?:conv=notrunc|seek=\d+|skip=\d+).*of=/dev/([hs]d[a-z]|mmcblk[0-9])'
  description: "Are you sure you want to write to a specific sector of the disk? This could corrupt data."
  id: fs:dd_advanced_disk_write
- from: fs
  test: 'gdisk\s+.*(/dev/([hs]d[a-z]|mmcblk[0-9]))'
  description: "Are you sure you want to modify GPT disk partitions? This could erase all data on the disk."
  id: fs:gdisk_disk_modify
- from: fs
  test: 'partprobe\s+.*(/dev/([hs]d[a-z]|mmcblk[0-9]))'
  description: "Are you sure you want to inform the OS of partition table changes? This could affect mounted partitions."
  id: fs:partprobe_disk_update
- from: fs
  test: 'blockdev\s+.*(/dev/([hs]d[a-z]|mmcblk[0-9]))'
  description: "Are you sure you want to modify block device parameters? This could affect disk operations."
  id: fs:blockdev_disk_modify
- from: fs
  test: 'mount\s+.*(/dev/([hs]d[a-z]|mmcblk[0-9]))'
  description: "Are you sure you want to mount this device? This could affect system stability."
  id: fs:mount_operations
- from: fs
  test: '(?:sudo\s+)?(?:lvremove|pvremove|vgremove)\s+(?:-[a-zA-Z]+\s+)?(?:--[a-zA-Z-]+\s+)?(?:/dev/(?:[hs]d[a-z]|mmcblk[0-9])|/dev/mapper/[^\s]+|[^\s]+)'
  description: "Are you sure you want to remove this logical volume/volume group? This will delete all data."
  id: fs:lvm_operations
- from: fs
  test: '(?:dump|restore)\s+.*(/dev/([hs]d[a-z]|mmcblk[0-9]))'
  description: "Are you sure you want to backup/restore this filesystem? This could affect system stability."
  id: fs:filesystem_backup
- from: fs
  test: 'cryptsetup\s+.*(/dev/([hs]d[a-z]|mmcblk[0-9]))'
  description: "Are you sure you want to encrypt/decrypt this device? This could affect data accessibility."
  id: fs:encryption_operations
