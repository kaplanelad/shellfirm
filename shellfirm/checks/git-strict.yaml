- from: git-strict
  test: git\s{1,}add\s{1,}.
  description: "Do you want to add all changes files to the staging area?"
  id: git-strict:add_all
- from: git-strict
  test: git\s{1,}commit\s{1,}(--all|-(.*a))
  description: "Do you want to commit all changes file and commit them?"
  id: git-strict:commit_all
- from: git-strict
  test: git\s{1,}rebase
  description: "This command will rewrite commit history by moving commits to a new base."
  id: git-strict:rebase
- from: git-strict
  test: git\s{1,}stash\s{1,}(pop|drop)
  description: "This command will apply or remove stashed changes."
  id: git-strict:stash_pop_drop
- from: git-strict
  test: git\s{1,}submodule\s{1,}(update|deinit)
  description: "This command will update or deinitialize Git submodules."
  id: git-strict:submodule_update
- from: git-strict
  test: git\s{1,}tag\s{1,}-a
  description: "This command will create an annotated tag."
  id: git-strict:create_tag
