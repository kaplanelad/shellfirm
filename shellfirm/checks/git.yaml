- from: git
  test: git\s{1,}reset
  description: "This command going to reset all your local changes."
  id: git:reset
- from: git
  test: git\s{1,}rm\s{1,}(\*|.)
  description: "This command going to delete all files."
  id: git:delete_all
- from: git
  test: git\s{1,}clean\s{1,}-fd
  description: "This command will remove all untracked files and directories."
  id: git:clean_force
- from: git
  test: git\s{1,}push\s{1,}(-f|--force)
  description: "This command will force push and overwrite remote history."
  id: git:force_push
- from: git
  test: git\s{1,}branch\s{1,}-D
  description: "This command will force delete a branch."
  id: git:force_delete_branch
- from: git
  test: git\s{1,}checkout\s{1,}-f
  description: "This command will force checkout and discard local changes."
  id: git:force_checkout
- from: git
  test: git\s{1,}rebase\s{1,}-i
  description: "This command will start an interactive rebase which can modify commit history."
  id: git:interactive_rebase
- from: git
  test: git\s{1,}filter-branch
  description: "This command will rewrite Git history and can be dangerous if used incorrectly."
  id: git:filter_branch
- from: git
  test: git\s{1,}gc\s{1,}--prune=now
  description: "This command will permanently delete unreachable objects."
  id: git:gc_prune
- from: git
  test: git\s{1,}update-ref\s{1,}-d
  description: "This command will delete a Git reference."
  id: git:delete_ref
- from: git
  test: git\s{1,}merge\s{1,}(--no-ff|--abort)
  description: "This command will perform a non-fast-forward merge or abort an ongoing merge."
  id: git:merge_no_ff
- from: git
  test: git\s{1,}cherry-pick
  description: "This command will apply changes from existing commits to the current branch."
  id: git:cherry_pick
- from: git
  test: git\s{1,}bisect
  description: "This command will start a binary search to find a commit that introduced a bug."
  id: git:bisect
- from: git
  test: git\s{1,}worktree\s{1,}(add|remove)
  description: "This command will add or remove a Git worktree."
  id: git:worktree_management
