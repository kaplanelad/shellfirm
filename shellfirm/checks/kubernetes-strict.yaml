- from: kubernetes-strict
  test: (k|kubectl)\s*delete\s*
  description: "This command will going to delete a given resource."
  id: kubernetes-strict:delete_resource
- from: kubernetes-strict
  test: ^(k|kubectl)\s*set\s*([[:word:]]+)
  description: "This command will update the given given resource."
  id: kubernetes-strict:update_resource
- from: kubernetes-strict
  test: (kubectl|k)\s*scale
  description: "This command will set a new size for a given resource."
  id: kubernetes-strict:change_resource_size
- from: kubernetes-strict
  test: (k|kubectl)\s*rollout\s*(pause|restart|resume|undo)
  description: "This command will manage a rollout for a given resource."
  id: kubernetes-strict:rollout_resource
