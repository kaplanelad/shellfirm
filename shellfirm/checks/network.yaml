- from: network
  test: iptables\s+-F
  description: "Flushing all firewall rules"
  id: network:flush_iptables
- from: network
  test: iptables\s+-X
  description: "Deleting all custom chains"
  id: network:delete_custom_chains
- from: network
  test: iptables\s+-t\s+nat\s+-F
  description: "Flushing all NAT rules"
  id: network:flush_nat_rules
- from: network
  test: ufw\s+disable
  description: "Disabling firewall"
  id: network:disable_firewall
- from: network
  test: ufw\s+--force\s+reset
  description: "Force resetting firewall rules"
  id: network:force_reset_firewall
- from: network
  test: systemctl\s+stop\s+networking
  description: "Stopping network service"
  id: network:stop_networking
- from: network
  test: systemctl\s+stop\s+NetworkManager
  description: "Stopping NetworkManager service"
  id: network:stop_network_manager
- from: network
  test: ifconfig\s+eth\d+\s+down
  description: "Bringing down network interface"
  id: network:bring_down_interface
- from: network
  test: ip\s+link\s+set\s+eth\d+\s+down
  description: "Bringing down network interface using ip command"
  id: network:bring_down_interface_ip
- from: network
  test: route\s+del\s+default
  description: "Deleting default route"
  id: network:delete_default_route
