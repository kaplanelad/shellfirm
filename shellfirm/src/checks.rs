//! Manage command checks

use std::{collections::HashMap, env};

use anyhow::Result;
use console::style;
use log::debug;
use rayon::prelude::*;
use regex::Regex;
use serde_derive::{Deserialize, Serialize};
use serde_regex;

use crate::{config::Challenge, prompt};

/// String with all checks from `checks` folder (prepared in build.rs) in YAML
/// format.
const ALL_CHECKS: &str = include_str!(concat!(env!("OUT_DIR"), "/all-checks.yaml"));

// list of custom filter
#[derive(Debug, Deserialize, Serialize, PartialEq, Eq, Hash, Clone)]
pub enum FilterType {
    IsExists,
    NotContains,
}

/// Describe single check
#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct Check {
    pub id: String,
    /// test ia a value that we check the command.
    #[serde(with = "serde_regex")]
    pub test: Regex,
    /// description of what is risky in this command
    pub description: String,
    /// the group of the check see files in `checks` folder
    pub from: String,
    #[serde(default)]
    pub challenge: Challenge,
    #[serde(default)]
    pub filters: HashMap<FilterType, String>,
}

/// Return all shellfirm check patterns
///
/// # Errors
/// when has an error when parsing check str to [`Check`] list
pub fn get_all() -> Result<Vec<Check>> {
    Ok(serde_yaml::from_str(ALL_CHECKS)?)
}

/// prompt a challenge to the user
///
/// # Errors
///
/// Will return `Err` when could not convert checks to yaml
pub fn challenge(
    challenge: &Challenge,
    checks: &[Check],
    deny_pattern_ids: &[String],
) -> Result<bool> {
    let mut descriptions: Vec<String> = Vec::new();
    let mut should_deny_command = false;

    debug!("list of denied pattern ids {:?}", deny_pattern_ids);

    for check in checks {
        if !descriptions.contains(&check.description) {
            descriptions.push(check.description.to_string());
        }
        if !should_deny_command && deny_pattern_ids.contains(&check.id) {
            should_deny_command = true;
        }
    }

    if should_deny_command {
        eprintln!("{}", style("##################").red().bold());
        eprintln!("{}", style("# COMMAND DENIED #").red().bold());
        eprintln!("{}", style("##################").red().bold());
    } else {
        eprintln!("{}", style("#######################").yellow().bold());
        eprintln!("{}", style("# RISKY COMMAND FOUND #").yellow().bold());
        eprintln!("{}", style("#######################").yellow().bold());
    }

    for description in descriptions {
        eprintln!("* {description}");
    }
    eprintln!();

    let show_challenge = challenge;
    if should_deny_command {
        debug!("command denied.");
        prompt::deny();
    }

    Ok(match show_challenge {
        Challenge::Math => prompt::math_challenge(),
        Challenge::Enter => prompt::enter_challenge(),
        Challenge::Yes => prompt::yes_challenge(),
    })
}

/// Check if the given command matched to on of the checks
///
/// # Arguments
///
/// * `checks` - List of checks that we want to validate.
/// * `command` - Command check.
#[must_use]
pub fn run_check_on_command(checks: &[Check], command: &str) -> Vec<Check> {
    checks
        .par_iter()
        .filter(|&v| v.test.is_match(command))
        .filter(|&v| check_custom_filter(v, command))
        .map(std::clone::Clone::clone)
        .collect()
}

/// filter custom checks
///
/// When true is returned it mean the filter should keep the check and not
/// filter our the check.
///
/// # Arguments
///
/// * `check` - Check struct
/// * `command` - Command.
fn check_custom_filter(check: &Check, command: &str) -> bool {
    if check.filters.is_empty() {
        return true;
    }
    // Capture command groups from the current check
    let caps = check.test.captures(command).unwrap();

    // by default true is return. it mean the check not filter out (safe side
    // security).
    let mut keep_check = true;
    for (filter_type, filter_params) in &check.filters {
        debug!(
            "filter information: command {} include filter: {:?} filter_params: {}",
            command, filter_type, filter_params
        );

        let keep_filter = match filter_type {
            FilterType::IsExists => filter_is_file_or_directory_exists(
                caps.get(filter_params.parse().unwrap())
                    .map_or("", |m| m.as_str()),
            ),
            FilterType::NotContains => filter_is_command_contains_string(command, filter_params),
        };

        if !keep_filter {
            keep_check = false;
            break;
        }
    }

    keep_check
}

/// check if the path exists (file and folder).
///
/// # Arguments
///
/// * `file_path` - check path.
fn filter_is_file_or_directory_exists(file_path: &str) -> bool {
    let mut file_path: String = file_path.trim().into();
    if file_path.starts_with('~') {
        match dirs::home_dir() {
            Some(path) => {
                file_path = file_path.replace('~', &path.display().to_string());
            }
            None => return true,
        };
    }

    if file_path.contains('*') {
        return true;
    }

    let full_path = match env::current_dir() {
        Ok(e) => e.join(file_path).display().to_string(),
        Err(err) => {
            log::debug!("could not get current dir. err: {:?}", err);
            return true;
        }
    };

    log::debug!("check is {} path is exists", full_path);
    std::path::Path::new(full_path.trim()).exists()
        || std::path::Path::new(full_path.trim()).is_dir()
}

fn filter_is_command_contains_string(command: &str, filter_params: &str) -> bool {
    !command.contains(filter_params)
}
#[cfg(test)]
mod test_checks {
    use std::fs;

    use insta::assert_debug_snapshot;
    use tempdir::TempDir;

    use super::*;

    const CHECKS: &str = r###"
- from: test-1
  test: test-(1)
  enable: true
  description: ""
  id: ""
- from: test-2
  test: test-(1|2)
  enable: true
  description: ""
  id: ""
- from: test-disabled
  test: test-disabled
  enable: true
  description: ""
  id: ""
"###;

    #[test]
    fn can_run_check_on_command() {
        let checks: Vec<Check> = serde_yaml::from_str(CHECKS).unwrap();
        assert_debug_snapshot!(run_check_on_command(&checks, "test-1"));
        assert_debug_snapshot!(run_check_on_command(&checks, "unknown command"));
    }

    #[test]
    fn can_check_custom_filter_with_file_exists() {
        let mut filters: HashMap<FilterType, String> = HashMap::new();
        filters.insert(FilterType::IsExists, "1".to_string());

        let check = Check {
            id: "id".to_string(),
            test: Regex::new(".*>(.*)").unwrap(),
            description: "some description".to_string(),
            from: "test".to_string(),
            challenge: Challenge::default(),
            filters,
        };

        let temp_dir = TempDir::new("config-app").unwrap();
        let app_path = temp_dir.path().join("app");
        fs::create_dir_all(&app_path).unwrap();
        let message_file = app_path.join("message.txt");

        let command = format!("cat 'write message' > {}", message_file.display());
        assert_debug_snapshot!(check_custom_filter(&check, command.as_ref()));
        std::fs::File::create(message_file).unwrap();
        assert_debug_snapshot!(check_custom_filter(&check, command.as_ref()));
    }

    #[test]
    fn can_check_custom_filter_with_str_contains() {
        let mut filters: HashMap<FilterType, String> = HashMap::new();
        filters.insert(FilterType::NotContains, "--dry-run".to_string());

        let check = Check {
            id: "id".to_string(),
            test: Regex::new("(delete)").unwrap(),
            description: "some description".to_string(),
            from: "test".to_string(),
            challenge: Challenge::default(),
            filters,
        };

        assert_debug_snapshot!(check_custom_filter(&check, "delete"));
        assert_debug_snapshot!(check_custom_filter(&check, "delete --dry-run"));
    }

    #[test]
    fn can_get_all_checks() {
        assert_debug_snapshot!(get_all().is_ok());
    }
}
