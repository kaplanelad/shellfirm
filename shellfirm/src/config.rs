//! Manage the app configuration by creating, deleting and modify the configuration

use crate::checks::Check;
use anyhow::anyhow;
use anyhow::Result as AnyResult;
use log::debug;
use serde_derive::{Deserialize, Serialize};
use std::fs;
use std::io;
use std::io::{Read, Write};
use std::time::{SystemTime, UNIX_EPOCH};

/// Default configuration file.
pub const DEFAULT_CONFIG_FILE: &str = include_str!("config.yaml");
/// String with all checks from `checks` folder (prepared in build.rs) in YAML format.
pub const ALL_CHECKS: &str = include_str!(concat!(env!("OUT_DIR"), "/all-checks.yaml"));

/// The method type go the check.
#[derive(Debug, Deserialize, Serialize, Clone, PartialEq)]
pub enum Method {
    /// Run start with check.
    StartWith,
    /// Run contains check.
    Contains,
    /// Run regex check.
    Regex,
}

/// The user challenge when user need to confirm the command.
#[derive(Debug, Deserialize, Serialize, Clone, PartialEq)]
pub enum Challenge {
    /// Math challenge.
    Math,
    /// Only enter will approve the command.
    Enter,
    /// only yes typing will approve the command.
    Yes,
    /// Default application challenge
    Default,
}

impl Default for Challenge {
    fn default() -> Self {
        Challenge::Default
    }
}

#[derive(Debug)]
/// describe configuration folder
pub struct Config {
    /// Configuration folder path.
    pub path: String,
    /// config file.
    pub config_file_path: String,
}

/// Describe the configuration yaml
#[derive(Debug, Deserialize, Serialize)]
pub struct Context {
    /// Type of the challenge.
    pub challenge: Challenge,
    /// List of all include files
    pub includes: Vec<String>,
    /// App version.
    #[serde(default)]
    pub version: String,
    /// List of checks.
    // #[serde(skip_deserializing)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub checks: Vec<Check>,
}

impl Config {
    /// Convert user config yaml to struct.
    ///
    /// # Errors
    ///
    /// Will return `Err` has an error when loading the config file
    pub fn load_config_from_file(&self) -> AnyResult<Context> {
        Ok(serde_yaml::from_str(&self.read_config_file()?)?)
    }

    /// Return default app config.
    /// # Errors
    ///
    /// Will return `Err` could not parse default config to yaml file
    pub fn load_default_config(&self) -> AnyResult<Context> {
        Ok(serde_yaml::from_str(DEFAULT_CONFIG_FILE)?)
    }

    /// update config file with the updated baseline checks.
    ///
    /// # Errors
    ///
    /// Will return `Err` adding check group return an error
    pub fn update_config_version(&self, config: &Context) -> AnyResult<()> {
        let mut config = self.add_checks_group(&config.includes)?;
        self.save_config_file_from_struct(&mut config)
    }

    /// Manage configuration folder & file.
    /// * Create config folder if not exists.
    /// * Create default config yaml file if not exists.
    ///
    /// # Errors
    ///
    /// Will return `Err` file could not created or loaded
    pub fn manage_config_file(&self) -> AnyResult<()> {
        self.create_config_folder()?;
        if fs::metadata(&self.config_file_path).is_err() {
            debug!("config file not found");
            self.create_default_config_file()?;
        }
        debug!("config content: {:?}", self.load_config_from_file()?);
        Ok(())
    }

    /// Update user settings files.
    ///
    /// # Arguments
    ///
    /// * `remove_checks` - if true the given `check_group` parameter will remove from configuration / if false will add.
    /// * `check_groups` - list of check groups to act.
    ///
    /// # Errors
    ///
    /// Will return `Err` group didn't added/removed
    pub fn update_config_content(
        &self,
        remove_checks: bool,
        check_groups: &[String],
    ) -> AnyResult<()> {
        if remove_checks {
            self.save_config_file_from_struct(&mut self.remove_checks_group(check_groups)?)?;
        } else {
            self.save_config_file_from_struct(&mut self.add_checks_group(check_groups)?)?;
        }
        Ok(())
    }

    /// Reset user configuration to the default app.
    ///
    /// # Errors
    ///
    /// Will return `Err` create config folder return an error
    pub fn reset_config(&self) -> AnyResult<()> {
        eprintln!(
            "Rest configuration will reset all checks settings. Select how to continue...\n \
            1. Yes, i want to override the current configuration\n \
            2. Override and backup the existing file\n \
            3. Cancel Or ^C"
        );
        let mut answer = String::new();
        io::stdin()
            .read_line(&mut answer)
            .expect("Failed to read line");

        match answer.trim() {
            "1" => self.create_default_config_file()?,
            "2" => {
                fs::rename(
                    &self.config_file_path,
                    format!(
                        "{}.{}.bak",
                        self.config_file_path,
                        SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs()
                    ),
                )?;
                self.create_default_config_file()?;
            }
            _ => return Err(anyhow!("unexpected option")),
        };

        Ok(())
    }

    /// Update default user challenge.
    ///
    /// # Arguments
    ///
    /// * `challenge` - new challenge to update
    ///
    /// # Errors
    ///
    /// Will return `Err` error return on load/save config
    pub fn update_challenge(&self, challenge: Challenge) -> AnyResult<()> {
        let mut conf = self.load_config_from_file()?;
        conf.challenge = challenge;
        self.save_config_file_from_struct(&mut conf)?;
        Ok(())
    }

    /// Create config folder if not exists.
    fn create_config_folder(&self) -> AnyResult<()> {
        if let Err(err) = fs::create_dir(&self.path) {
            if err.kind() != std::io::ErrorKind::AlreadyExists {
                return Err(anyhow!("could not create folder: {}", err));
            }
            debug!("configuration folder found: {}", &self.path);
        } else {
            debug!("configuration created in path: {}", &self.path);
        }
        Ok(())
    }

    /// Create config file from default template.
    fn create_default_config_file(&self) -> AnyResult<()> {
        let mut conf = self.load_default_config()?;
        conf.checks = self.get_default_checks(&conf.includes)?;
        self.save_config_file_from_struct(&mut conf)
    }

    /// Convert the given config to YAML format and the file.
    ///
    /// # Arguments
    ///
    /// * `config` - Config struct
    fn save_config_file_from_struct(&self, mut config: &mut Context) -> AnyResult<()> {
        config.version = env!("CARGO_PKG_VERSION").to_string();
        let content = serde_yaml::to_string(config)?;
        let mut file = fs::File::create(&self.config_file_path)?;
        file.write_all(content.as_bytes())?;
        debug!(
            "config file crated in path: {}. config data: {:?}",
            &self.config_file_path, config
        );
        Ok(())
    }

    /// Return config content.
    fn read_config_file(&self) -> AnyResult<String> {
        let mut file = std::fs::File::open(&self.config_file_path)?;
        let mut content = String::new();
        file.read_to_string(&mut content)?;
        Ok(content)
    }

    /// Add checks group to user configuration.
    ///
    /// # Arguments
    ///
    /// * `checks_group` - list of groups to add.
    fn add_checks_group(&self, checks_group: &[String]) -> AnyResult<Context> {
        //load user config file
        match self.load_config_from_file() {
            Ok(mut conf) => {

                for c in checks_group{
                    if !conf.includes.contains(c){
                        conf.includes.push(c.clone());
                    }
                }
                debug!("new list of includes groups: {:?}", checks_group);

                // List of checks that the user disable or change the challenge type
                let override_check_settings = conf.checks.iter()
                    .filter(|&c| checks_group.contains(&c.from))
                    .filter(|c| !c.enable || c.challenge != Challenge::Default)
                    .cloned()
                    .collect::<Vec<Check>>();

                debug!("override checks settings: {:?}", override_check_settings);

                // remove checks group that we want to add for make sure that we not have duplicated checks
                let mut checks = conf.checks.iter()
                    .filter(|&c| !checks_group.contains(&c.from))
                    .cloned()
                    .collect::<Vec<Check>>();

                checks.extend( self.get_default_checks(checks_group)?);

                for override_check in override_check_settings{
                    for c in  &mut checks{
                        if c.test == override_check.test{
                            c.enable = override_check.enable;
                            c.challenge = override_check.challenge.clone();
                        }
                    }
                }

                conf.checks = checks;
                debug!("new check list: {:?}", conf.checks);
                Ok(conf)
            },
            Err(e) => return Err(anyhow!("could not parse current config file. please try to fix the yaml. Try resolving by running `shellfirm config reset` Error: {}", e))
        }
    }

    /// Remove checks group from user configuration
    ///
    /// # Arguments
    ///
    /// * `checks_group` - list of groups to add.
    fn remove_checks_group(&self, checks_group: &[String]) -> AnyResult<Context> {
        //load user config file
        match self.load_config_from_file() {
            Ok(mut conf) => {

                for c in checks_group{
                    if conf.includes.contains(c){
                        conf.includes.retain(|x| x != c);
                    }
                }
                debug!("new list of includes groups: {:?}", checks_group);
                // remove checks group that we want to add for make sure that we not have duplicated checks
                conf.checks = conf.checks.iter().filter(|&c| conf.includes.contains(&c.from)).cloned().collect::<Vec<Check>>();

                debug!("new check list: {:?}", conf.checks);
                Ok(conf)
            },
            Err(_e) => return Err(anyhow!("could not parse current config file. please try to fix the yaml file or override the current configuration by use the flag `--behavior override`"))
        }
    }

    fn get_default_checks(&self, includes: &[String]) -> AnyResult<Vec<Check>> {
        Ok(get_all_available_checks()?
            .iter()
            .filter(|&c| includes.contains(&c.from))
            .cloned()
            .collect::<Vec<Check>>())
    }
}

/// Get application  setting config.
///
/// # Errors
///
/// Will return `Err` error return on load/save config
pub fn get_config_folder() -> AnyResult<Config> {
    let package_name = env!("CARGO_PKG_NAME");

    match dirs::home_dir() {
        Some(path) => {
            let config_folder = path.join(format!(".{}", package_name));

            let setting_config = Config {
                path: config_folder.to_str().unwrap_or("").to_string(),
                config_file_path: config_folder
                    .join("config.yaml")
                    .to_str()
                    .unwrap_or("")
                    .to_string(),
            };

            setting_config.manage_config_file()?;
            debug!("configuration settings: {:?}", setting_config);
            Ok(setting_config)
        }
        None => return Err(anyhow!("could not get directory path")),
    }
}

/// parse `ALL_CHECKS` const to vector of checks
fn get_all_available_checks() -> AnyResult<Vec<Check>> {
    Ok(serde_yaml::from_str(ALL_CHECKS)?)
}

#[cfg(test)]
mod test_config {
    use super::*;
    use insta::assert_debug_snapshot;
    use std::path::Path;

    fn get_current_project_path() -> String {
        std::env::current_dir().unwrap().to_str().unwrap().into()
    }

    fn get_temp_config_folder(file_name: &str) -> Config {
        let tmp_folder = Path::new(&get_current_project_path())
            .join("tmp")
            .to_str()
            .unwrap()
            .to_string();
        let config_file_path = Path::new(&tmp_folder)
            .join(file_name)
            .to_str()
            .unwrap()
            .to_string();
        if fs::metadata(&config_file_path).is_ok() {
            fs::remove_file(&config_file_path).unwrap();
        }

        Config {
            path: tmp_folder,
            config_file_path,
        }
    }

    #[test]
    fn can_load_config_from_file() {
        let settings_config = Config {
            path: get_current_project_path(),
            config_file_path: Path::new(&get_current_project_path())
                .join("src")
                .join("config.yaml")
                .to_str()
                .unwrap()
                .to_string(),
        };

        assert_debug_snapshot!(settings_config.load_config_from_file());
    }

    #[test]
    fn can_write_config_file() {
        let settings_config = get_temp_config_folder("config.yaml");
        assert_debug_snapshot!(settings_config.manage_config_file());
    }

    #[test]
    fn can_create_default_config_file() {
        let settings_config = get_temp_config_folder("default.yaml");
        assert_debug_snapshot!(settings_config.create_default_config_file());
        assert_debug_snapshot!(Path::new(&settings_config.config_file_path).exists());
    }

    #[test]
    fn can_save_config_file_from_struct() {
        let settings_config = get_temp_config_folder("save-from-struct.yaml");

        let mut config = settings_config.load_default_config().unwrap();

        // creates configuration file with only 1 check (we want to check that we not change existing check and just append new ones)
        config.checks = vec![Check {
            test: String::from("is value"),
            method: Method::Contains,
            enable: true,
            description: String::from("description"),
            from: String::from("from"),
            challenge: Challenge::Default,
            filters: std::collections::HashMap::new(),
        }];

        assert_debug_snapshot!(settings_config.save_config_file_from_struct(&mut config));
        assert_debug_snapshot!(settings_config.load_config_from_file());
    }

    #[test]
    fn can_add_checks_group() {
        let settings_config = get_temp_config_folder("add-checks.yaml");

        // let mut config = settings_config.load_default_config().unwrap();
        let mut config = Context {
            challenge: Challenge::Default,
            includes: vec!["test".to_string()],
            version: "0.0.1".to_string(),
            checks: vec![
                Check {
                    test: String::from("test-value"),
                    method: Method::Contains,
                    enable: true,
                    description: String::from("description"),
                    from: String::from("test"),
                    challenge: Challenge::Default,
                    filters: std::collections::HashMap::new(),
                },
                Check {
                    test: String::from("test-1value"),
                    method: Method::Contains,
                    enable: true,
                    description: String::from("description"),
                    from: String::from("test1"),
                    challenge: Challenge::Default,
                    filters: std::collections::HashMap::new(),
                },
                Check {
                    test: String::from("test2-value"),
                    method: Method::Contains,
                    enable: true,
                    description: String::from("description"),
                    from: String::from("test2"),
                    challenge: Challenge::Default,
                    filters: std::collections::HashMap::new(),
                },
            ],
        };

        assert_debug_snapshot!(settings_config.save_config_file_from_struct(&mut config));
        assert_debug_snapshot!(settings_config.add_checks_group(&["test2".to_string()]));
    }

    // #[ignore]
    // #[test]
    // fn can_remove_checks_group() {
    //     let settings_config = get_temp_config_folder("add-checks.yaml").unwrap();

    //     let mut config = settings_config.load_default_config().unwrap();

    //     config.includes = vec!["test".into()];
    //     config.checks = vec![Check {
    //         test: String::from("is value"),
    //         method: Method::Contains,
    //         enable: true,
    //         description: String::from("description"),
    //         from: String::from("test"),
    //         challenge: Challenge::Default,
    //         filters: std::collections::HashMap::new(),
    //     }];

    //     assert_debug_snapshot!(settings_config.save_config_file_from_struct(&mut config));
    //     assert_debug_snapshot!(settings_config.remove_checks_group(&["test".into()]));
    // }
}
