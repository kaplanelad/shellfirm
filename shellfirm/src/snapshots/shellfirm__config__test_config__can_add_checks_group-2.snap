---
source: shellfirm/src/config.rs
expression: "settings_config.add_checks_group(&[\"test2\".to_string()])"
---
Ok(
    Context {
        challenge: Default,
        includes: [
            "test",
            "test2",
        ],
        version: "0.2.2",
        checks: [
            Check {
                test: "test-value",
                method: Contains,
                enable: true,
                description: "description",
                from: "test",
                challenge: Default,
                filters: {},
            },
            Check {
                test: "test-1value",
                method: Contains,
                enable: true,
                description: "description",
                from: "test1",
                challenge: Default,
                filters: {},
            },
        ],
    },
)
